### PR TITLE
Add `enum_case_associated_values_count` opt-in rule

### PR DESCRIPTION
**Pull request by lakpa**
Resolves [2997](https://github.com/realm/SwiftLint/issues/2997)

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+        var violations: [StyleViolation] = []
+
+        for element in enumElements {
+            guard let keyOffset = element.offset,
+                let keyName = element.name else {
+                continue
+            }
+            let enumCaseAssociatedValueCount = keyName.split(separator: ":").count - 1
+            if enumCaseAssociatedValueCount >= configuration.warning {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift]: You should validate the error threshold as well, check [FileLengthRule](https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Rules/Metrics/FileLengthRule.swift#L40) for an example.

It'd be nice to have a better message as well, like `"Enum case \(caseNameWithoutParams) should contain \(parameter.value) associated values or less: currently contains \(enumCaseAssociatedValueCount)"`

File: Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            """
+        ]
+    )
+
+    public func validate(
+        file: SwiftLintFile,
+        kind: SwiftDeclarationKind,
+        dictionary: SourceKittenDictionary
+    ) -> [StyleViolation] {
+        guard kind == .enum else {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift]: could you check for `enumElement` here and remove the substructure navigation below? this makes the rule more focused and leaves the navigation to `ASTRule`

File: Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            enum Employee {
+                case fullTime(name: String, retirement: Date, designation: String, contactNumber: Int)
+                case partTime(name: String, age: Int, contractEndDate: Date)
+            }
+            """
+        ],
+        triggeringExamples: [
+            """
+            enum Employee {
+                case fullTime(name: String, retirement: Date, age: Int, designation: String, contactNumber: Int)
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift]: Can you please add the violation mark (`↓`) in the examples so we can test that the violation is triggered in the expected location?

It'd be great to also add examples with unnamed associated values (e.g. `bar(String)`)

File: Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+public struct EnumCaseAssociatedValuesLengthRule: ASTRule, OptInRule, ConfigurationProviderRule {
+    public var configuration = SeverityLevelsConfiguration(warning: 5, error: 6)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "enum_case_associated_values_count",
+        name: "Enum Case Associated Values Count",
+        description: "Number of associated values should be low",
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift]: maybe elaborate the description a bit more to mention that this is related to enum case declarations?

**Comment by lakpa** [comment]
@marcelofabri addressed your comments, please review

File: Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+public struct EnumCaseAssociatedValuesLengthRule: ASTRule, OptInRule, ConfigurationProviderRule {
+    public var configuration = SeverityLevelsConfiguration(warning: 5, error: 6)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "enum_case_associated_values_count",
+        name: "Enum Case Associated Values Count",
+        description: "Enum case's associated value length should be low",
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Metrics/EnumCaseAssociatedValuesLengthRule.swift]: can you replace "length" with "count" or use something like "Number of associated values in an enum case should be low."

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,7 +10,9 @@
 
 #### Enhancements
 
-* None.
+* Add `enum_case_associated_value_count` opt-in rule.  
+  [lakpa](https://github.com/lakpa)
+  [#issue_number](https://github.com/realm/SwiftLint/pull/2998)
```
Comment (marcelofabri) [inline CHANGELOG.md]: this should be referencing https://github.com/realm/SwiftLint/issues/2997

**Comment by marcelofabri** [comment]
@lakpa Thanks for the changes! I've rebased to fix a conflict and fixed the changelog entry position that was wrong after the rebase. Sorry for the force push! I'll merge it once CI is green ✅

**Pull request merged**